octopusdeploy_framework/datasource_tenant_projects.go

```diff
@@ -0,0 +1,201 @@
+package octopusdeploy_framework
+
+import (
+	"context"
+	"fmt"
+	"github.com/OctopusDeploy/go-octopusdeploy/v2/pkg/client"
+	"github.com/OctopusDeploy/go-octopusdeploy/v2/pkg/tenants"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/schemas"
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/util"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	datasourceSchema "github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"golang.org/x/exp/slices"
+)
+
+type TenantProjectsDataModel struct {
+	SpaceID        types.String `tfsdk:"space_id"`
+	TenantIDs      types.List   `tfsdk:"tenant_ids"`
+	ProjectIDs     types.List   `tfsdk:"project_ids"`
+	EnvironmentIDs types.List   `tfsdk:"environment_ids"`
+	TenantProjects types.List   `tfsdk:"tenant_projects"`
+}
+
+type tenantProjectsDataSource struct {
+	*Config
+}
+
+func NewTenantProjectDataSource() datasource.DataSource {
+	return &tenantProjectsDataSource{}
+}
+
+func (*tenantProjectsDataSource) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = util.GetTypeName("tenant_projects")
+}
+
+func (t *tenantProjectsDataSource) Configure(_ context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	t.Config = DataSourceConfiguration(req, resp)
+}
+
+func (*tenantProjectsDataSource) Schema(_ context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = datasourceSchema.Schema{
+		Description: "Provides information about existing tenants.",
+		Attributes: map[string]datasourceSchema.Attribute{
+			"tenant_ids":      schemas.GetQueryIDsDatasourceSchema(),
+			"project_ids":     schemas.GetQueryIDsDatasourceSchema(),
+			"environment_ids": schemas.GetQueryIDsDatasourceSchema(),
```

Comment:
I can see a case where customers may want to get a list of all the TenantProject relationships for a given environment, but the current implementation starts always from a Project and/or Tenant.

But I also appreciate that the current shape of the public API surface would make this quite expensive/slow to do. You'd need to start from the all projects endpoint, and then filter by environment in-memory, and then query for tenants by project ID. This seems both a trap for poor performance, and much more complicated than the logic for a datasource should be.

There is a BFF endpoint which is used by the Tenants Overview page in the UI, which allows the kind of searching that would enable us to do this in a performant way. 

I'd suggest we proceed with this first iteration maintaining the restriction of needing to be based on ProjectIDs or TenantIDs with Environment just as a results filter. We can separately prioritise upgrading that BFF to a public endpoint, at which point we could unlock "stage two" of this datasource enabling search by any combination of the inputs.

Reply:
I've raised [🔒 an internal ticket](https://app.shortcut.com/octopusdeploy/story/89475/upgrade-tenantsoverview-bff-to-the-public-api-surface) to promote the Tenants Overview BFF to the public API surface.